src/EditorFeatures/Core/Remote/SolutionChecksumUpdater.cs

```diff
@@ -21,115 +19,148 @@ namespace Microsoft.CodeAnalysis.Remote
     {
         private readonly Workspace _workspace;
-        private readonly TaskQueue _textChangeQueue;
-        private readonly AsyncQueue<IAsyncToken> _workQueue = new();
-        private readonly object _gate = new();
-
-        private CancellationTokenSource _globalOperationCancellationSource;
-
-        // hold the async token from WaitAsync so ExecuteAsync can complete it
-        private IAsyncToken _currentToken;
-
-        public SolutionChecksumUpdater(Workspace workspace, IGlobalOptionService globalOptions, IAsynchronousOperationListenerProvider listenerProvider, CancellationToken shutdownToken)
-            : base(listenerProvider.GetListener(FeatureAttribute.SolutionChecksumUpdater),
-                   workspace.Services.GetService<IGlobalOperationNotificationService>(),
-                   TimeSpan.FromMilliseconds(globalOptions.GetOption(RemoteHostOptions.SolutionChecksumMonitorBackOffTimeSpanInMS)), shutdownToken)
+        private readonly IGlobalOperationNotificationService _globalOperationService;
+
+        /// <summary>
+        /// Queue to push out text changes in a batched fashion when we hear about them.  Because these should be short
+        /// operations (only syncing text changes) we don't cancel this when we enter the paused state.  We simply don't
+        /// start queuing more requests into this until we become unpaused.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue<(Document? oldDocument, Document? newDocument)> _textChangeQueue;
+
+        /// <summary>
+        /// Queue for kicking off the work to synchronize the primary workspace's solution.  The cancellation token is
+        /// used so that we can stop the work when we enter the paused state.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue<CancellationToken> _synchronizeWorkspaceQueue;
+
+        /// <summary>
+        /// Cancellation series we use to stop work when we get paused.
+        /// </summary>
+        private readonly CancellationSeries _pauseCancellationSeries;
+
+        /// <summary>
+        /// Cancellation token we trigger when we enter the paused state.
+        /// </summary>
+        private CancellationToken _currentWorkToken;
+
+        public SolutionChecksumUpdater(
+            Workspace workspace,
+            IAsynchronousOperationListenerProvider listenerProvider,
+            CancellationToken shutdownToken)
         {
+            var listener = listenerProvider.GetListener(FeatureAttribute.SolutionChecksumUpdater);
+            _globalOperationService = workspace.Services.GetRequiredService<IGlobalOperationNotificationService>();
+
+            _pauseCancellationSeries = new CancellationSeries(shutdownToken);
             _workspace = workspace;
-            _textChangeQueue = new TaskQueue(Listener, TaskScheduler.Default);
+
+            _textChangeQueue = new AsyncBatchingWorkQueue<(Document? oldDocument, Document? newDocument)>(
```

Comment:
So I see we have two queues because there are two different calls to synchronize the primary workspace and synchronize text changes.  So question...

My naive implementation would be to just have one call that says 'SynchronizePrimaryWorkspace' that gives it the tree? of checksums, which the remote host can then diff against what it has and callback asking for what it is missing.  However I'm guessing that could be expensive in the common case when one document is just changing over and over.  Is that roughly a correct understanding?

Reply:
Talked offline.  The reason for the text-syncing is to optimize teh common case of lots of edits happening to a file.  We can just preemptively populate oop with teh contents of those docs just by sending the text changes over.